src/internal/shoelace-element.ts

```diff
@@ -1,13 +1,19 @@
 import { LitElement } from 'lit';
 import { property } from 'lit/decorators.js';
 
+type SlEventInit<T> = T extends keyof GlobalEventHandlersEventMap
+  ? GlobalEventHandlersEventMap[T] extends CustomEvent
+    ? CustomEventInit<GlobalEventHandlersEventMap[T]['detail']>
+    : CustomEventInit
+  : CustomEventInit;
+
 export default class ShoelaceElement extends LitElement {
   // Make localization attributes reactive
   @property() dir: string;
   @property() lang: string;
 
   /** Emits a custom event with more convenient defaults. */
-  emit(name: string, options?: CustomEventInit) {
+  emit<T extends string>(name: T, options?: SlEventInit<T>) {
```

Comment:
This ensures that if you pass in a detail object for an event, it must match what that event expects for its detail. It never requires you to pass in a details object though, so it won't complain if you forget to pass in even details for an event that should require it. I haven't figured out a way to get that to work.

Reply:
I'm OK with that. Might be worth a `// TODO` though.

Reply:
I've improved it to no longer allow an `options` object without a `detail` property if the event has details, but you can still omit the `options` parameter entirely. I'm not sure if its possible to make a parameter conditionally optional in typescript. Added a `// TODO`.

Reply:
I dig some more investigation and it turns out that Typescript actually supports function overloads, which is news to me. With a little metaprogramming, I was able to get it to work, and `emit` is now fully type safe. If you attempt to call `emit` with an event that has detail (and at least one of the detail properties is non-optional), it will now error if you don't provide the event detail.

Reply:
Amazing. Thanks for the update!